design/secret_rotation_design.md

```diff
@@ -0,0 +1,951 @@
+    (in `file` mode) when secret values have changed.
+
+    The periodic secret refresh interval is configurable. The default refresh
+    interval is 5 minutes.
+
+    Note that for `k8s_secrets` mode, updates to Kubernetes Secrets will
+    only have an immediate effect if the application is consuming Kubernetes
+    Secrets values via volume mounts (as opposed to consuming Kubernetes
+    Secrets via environment variables).
+
+    **User Stories:** US-03, US-04
+
+  - **No Updates if Content Remains Unchanged**
+
+    Kubernetes Secrets (in `k8s_secrets` mode) and secret files (in `file`
+    mode) are NOT updated if their content remains unchanged. This will allow
+    applications to view file metadata (e.g. timestamp) to determine if
+    secret files have been updated.
+
+    **User Story:** US-03
+
+  - **Secrets/Files Sanitized When Secrets Deleted/Revoked From Conjur**
+    
+    When one or more secrets is deleted, or access to one or more secrets is
+    revoked from Conjur, then the Secrets Provider will:
+
+    - **Remove the associated secret values from Kubernetes Secrets.**
+      (for `k8s-secret` destination mode)
+    - **Delete any secret files that contain the associated secret values.**
+      (for `file` mode)
+    - **Exit with error.** The Secrets Provider treats this as a fatal error
+      whether or not this occurs during initial startup or as part of
+      periodic secret refresh.
+
+    **User Stories:** US-02, US-04
+
+  - **Atomic Writes for Secret Files**
+
+    When each secret file is written (in `file` mode), the write operation is
+    atomic. This will ensure that the application will always read a valid
+    instance of each secret file even though it is reading the file
+    asynchronously with respect to the file being written by the Secrets
+    Provider.
+
+    **User Story:** US-01
+
+  - **Secret/File Updates for Changes Other than Conjur Secrets:**
+
+    There are several components that are used by the Secrets Provider in
+    rendering Kubernetes Secret or secret file content. Changes to any of these
+    components may result in changes to the rendered content. For example,
+    changes to the following will trigger the Secrets Provider to update
+    Kubernetes Secrets and secret file content:
+
+    - Changes to a custom template that is configured via a ConfigMap (`file` destination mode)
+    - Changes to the data field of a source Kubernetes Secret (`k8s-secret` destination mode)
+
+    **User Story:** US-03
+    
+  - **Use of SHA-256 checksums to Detect Changes:**<br />
+    For added security, the Secrets Provider will not cache any unencrypted,
+    retrieved secret values in order to detect when Kubernetes Secrets or
+    Push-to-File secret files need to be updated. Instead, the Secrets Provider
+    will store a SHA-256 checksum of the last rendered content that was written
+    to each target Kubernetes Secret or secret file for detecting when contents
+    have changed.
```

Comment:
What happens if the file is deleted by the application?  Will SP restore the file @ next sync?  Will SP compare against the cached SHA-256 checksum so file will not be re-created if checksum matches and will only be re-created if checksum does not match?